exchanges/kline/builder.go

```diff
@@ -0,0 +1,139 @@
+package kline
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+var (
+	// ErrUnsetName is an error for when the exchange name is not set
+	ErrUnsetName  = errors.New("unset exchange name")
+	errNilBuilder = errors.New("nil kline builder")
+)
+
+// Builder is a helper to request and convert time series to a required candle
+// interval.
+type Builder struct {
+	Name      string
+	Pair      currency.Pair
+	Formatted currency.Pair
+	Asset     asset.Item
+	Request   Interval
+	Required  Interval
+	Start     time.Time
+	End       time.Time
+}
+
+// GetBuilder generates a builder for interval conversions supported by an
+// exchange.
+func GetBuilder(name string, pair, formatted currency.Pair, a asset.Item, required, request Interval, start, end time.Time) (*Builder, error) {
+	if name == "" {
+		return nil, ErrUnsetName
+	}
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if formatted.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if !a.IsValid() {
+		return nil, asset.ErrNotSupported
+	}
+	if required == 0 {
+		return nil, fmt.Errorf("required %w", ErrUnsetInterval)
+	}
+	if request == 0 {
+		return nil, fmt.Errorf("request %w", ErrUnsetInterval)
+	}
+	err := common.StartEndTimeCheck(start, end)
+	if err != nil {
+		return nil, err
+	}
+	// Force alignment to request interval
+	start = start.Truncate(request.Duration())
+	end = end.Truncate(request.Duration())
+	return &Builder{name, pair, formatted, a, request, required, start, end}, nil
+}
+
+// GetRanges returns the date ranges for candle intervals broken up over
+// requests
+func (b *Builder) GetRanges(limit uint32) (*IntervalRangeHolder, error) {
+	if b == nil {
+		return nil, errNilBuilder
+	}
+	return CalculateCandleDateRanges(b.Start, b.End, b.Request, limit)
+}
+
+// ConvertCandles converts time series candles into a kline.Item type. This will
+// auto convert from a lower to higher time series if applicable.
+func (b *Builder) ConvertCandles(timeSeries []Candle) (*Item, error) {
+	if b == nil {
+		return nil, errNilBuilder
+	}
+
+	holder := &Item{
+		Exchange: b.Name,
+		Pair:     b.Pair,
+		Asset:    b.Asset,
+		Interval: b.Request,
+		Candles:  timeSeries,
+	}
+
+	// NOTE: timeSeries param above must keep underlying slice reference in this
+	// function as it is used for method ConvertCandles on type BuilderExtended
+	// for SetHasDataFromCandles candle matching.
+	// TODO: Shift burden of proof to the caller e.g. only find duplicates and error.
```

Comment:
One of the main reasons why I have pushed for this kind of shift is the `orderbook package`. Removing the filter/repair layer exposed a lot of implementation and data integrity flaws. There was inconsistencies throughout most of these implementations anyway; time window checking etc. I have funneled that through one base method call to reduce these implementation complexities. As with the orderbook caveats this can be handled in wrapper. As for now I have tried to optimize and standardize as much as I can before that change occurs. I want to make sure that if there are inconsistencies we catch them, expose them and program more efficiently.

Reply:
This is the kind of stuff I would love to read in a PR description so I can understand your intent and future plans so I can keep those in mind when making comments